parts/k8s/kuberneteswindowssetup.ps1

```diff
@@ -192,7 +200,13 @@ try
             -UserAssignedClientID $global:UserAssignedClientID `
             -UseInstanceMetadata $global:UseInstanceMetadata `
             -LoadBalancerSku $global:LoadBalancerSku `
-            -ExcludeMasterFromStandardLB $global:ExcludeMasterFromStandardLB
+            -ExcludeMasterFromStandardLB $global:ExcludeMasterFromStandardLB `
+            -TargetEnvironment $TargetEnvironment
+
+        {{if IsAzureStackCloud}}
+        $azureStackConfigFile = [io.path]::Combine($global:KubeDir, "azurestackcloud.json")
+        "{{ GetEscapedEnvironmentJSON }}" | Out-File -encoding ASCII -filepath "$azureStackConfigFile"

```

Comment:
Can you share what the contents of this file are? Since this file will be included in the contents of a custom script extension, that means it has to be escaped for both powershell and ARM. Single quotes break ARM

My preference would be to split this up into separate variables up at the top, then have function like `Write-AzureConfig` above that handles creating the file. I'm planning to delete kuberneteswindowssetup.ps1 as it exists today. If that code is in another function we can still reuse it after the refactor.

Reply:
The list of "stuff" needed to configure a VM will be going into an API or config file, and the code applying the config will be put in the VM image instead of being passed through a custom script extension.

Reply:
```
{
    "name": "AzureStackCloud",
    "managementPortalURL": "https://portal.local.azurestack.external/",
    "publishSettingsURL": "",
    "serviceManagementEndpoint": "https://management.azurestack.onmicrosoft.com/00000000-0000-0000-0000-0000000000",
    "resourceManagerEndpoint": "https://management.local.azurestack.external/",
    "activeDirectoryEndpoint": "https://login.microsoftonline.com/",
    "galleryEndpoint": "https://galleryartifacts.hosting.local.azurestack.external/galleryartifacts/",
    "keyVaultEndpoint": "",
    "graphEndpoint": "https://graph.windows.net/",
    "serviceBusEndpoint": "",
    "batchManagementEndpoint": "",
    "storageEndpointSuffix": "local.azurestack.external",
    "sqlDatabaseDNSSuffix": "",
    "trafficManagerDNSSuffix": "",
    "keyVaultDNSSuffix": "vault.local.azurestack.external",
    "serviceBusEndpointSuffix": "",
    "serviceManagementVMDNSSuffix": "",
    "resourceManagerVMDNSSuffix": "cloudapp.azurestack.external",
    "containerRegistryDNSSuffix": "",
    "tokenAudience": ""
}
```

Reply:
```
{
    "name": "AzureStackCloud",
    "managementPortalURL": "https://portal.local.azurestack.external/",
    "publishSettingsURL": "",
    "serviceManagementEndpoint": "https://management.azurestack.onmicrosoft.com/00000000-0000-0000-0000-0000000000",
    "resourceManagerEndpoint": "https://management.local.azurestack.external/",
    "activeDirectoryEndpoint": "https://login.microsoftonline.com/",
    "galleryEndpoint": "https://galleryartifacts.hosting.local.azurestack.external/galleryartifacts/",
    "keyVaultEndpoint": "",
    "graphEndpoint": "https://graph.windows.net/",
    "serviceBusEndpoint": "",
    "batchManagementEndpoint": "",
    "storageEndpointSuffix": "local.azurestack.external",
    "sqlDatabaseDNSSuffix": "",
    "trafficManagerDNSSuffix": "",
    "keyVaultDNSSuffix": "vault.local.azurestack.external",
    "serviceBusEndpointSuffix": "",
    "serviceManagementVMDNSSuffix": "",
    "resourceManagerVMDNSSuffix": "cloudapp.azurestack.external",
    "containerRegistryDNSSuffix": "",
    "tokenAudience": ""
}
```

Reply:
^^^ Here is an example of the file as it would be found on the machine. 
The file is different depending on which azure stack instance the user is trying to connect to, so it can't be on the VM image.

Reply:
There is an example of the escaped file in the unit test https://github.com/Azure/aks-engine/blob/2e50b6343b7775e5da87c0c6b7be518c04dcb598/pkg/engine/template_generator_test.go#L175

I escape the `"` with `""` in this file https://github.com/Azure/aks-engine/blob/2e50b6343b7775e5da87c0c6b7be518c04dcb598/pkg/engine/template_generator.go#L626

I should probably escape single quotes as well so that if the input has them they don't break the template. 

I could make separate variables at the top, but there would be around 20 of them given the file structure above. Should I add all 20 variables?

Reply:
Also, with linux we just use cloud-init to place the file on the host during provisioning. Is there a way to add multiple custom data files to a windows host while provisioning?

Reply:
unfortunately there isn't a way to add multiple files during provisioning, they need to be written from the PowerShell scripts

Reply:
That's too bad. So would you rather I put it in a function with a bunch of new variables or pass the escaped string?

Reply:
Initially I thought some of those might be needed elsewhere, but it looks like that JSON never really needs to be parsed as input, just placed. From that perspective I think we should just get it onto disk rather than passing it across functions.

What I was initially asking would mean you couldn't reuse this between Linux & Windows nodes. We shouldn't reimplement it in PowerShell.
https://github.com/azure/aks-engine/blob/c68c2a94b63ff401241b88fbca104f000a9481e7/pkg/api/types.go#L1698

I think the cleanest thing here, and the closest to dropping a file with cloud-init is to:

1. Change template_generator.go to use `base64.StdEncoding.EncodeToString`  to encode the output from `GetCustomEnvironmentJSON`

2. write it out like this https://github.com/azure/aks-engine/blob/407207ba834c2fea9ea93d8a23d36f67519e2f96/parts/k8s/kuberneteswindowssetup.ps1#L119

Reply:
base64 avoids all the differences in escapes. once that's done it looks ready to merge